File: leetcode/tree/124. Binary Tree Maximum Path Sum/js/main.js
// Time: O(N)
// Space: Мы должны сохранить рекурсивный стек размером с высоту дерева, который равен O (log N) для двоичного дерева.
const maxPathSum = root => {
  let ans = -Number.MAX_VALUE;

  // Bottom-top рекурсия

  dfs(root);

  return ans;

  // На каждом узле нам необходимо принять решение: если сумма левого пути, большего, правого пути
  // Мы выбираем левый путь и добавляем значение текущего узла, эта рекурсия проходит вплоть до корневого узла.
  function dfs(node) {
    if (node == null) {
      return 0;
    }

    let l = dfs(node.left);
    let r = dfs(node.right);

    ans = Math.max(ans, l + node.val, r + node.val, l + r + node.val, node.val);

    return Math.max(l + node.val, r + node.val, 0);
  }
};

File: leetcode/tree/124. Binary Tree Maximum Path Sum/py/main.py
class Solution:
    # Time O(N)
    # Space O(N)
    def maxPathSum(self, root: TreeNode) -> int:
        def dfs(node):
            if node is None:
                return 0

            L = dfs(node.left)
            R = dfs(node.right)

            current = 0

            if L > 0:
                current += L

            if R > 0:
                current += R

            self.ans = max(self.ans, node.val + current)

            return node.val + max(L, R, 0)

        self.ans = -float('inf')
        dfs(root)

        return self.ans
